### PR TITLE
feat: allow Spanner and Datastore Transaction Manager in same project

### LINKED ISSUE #944: Transaction error using Spanner JDBC with JPA and Spring Cloud GCP Datastore in the same project
Hello.

I'm trying to configure a single project using Spring Data JPA + Spanner using JDBC connector and also the Spring Cloud GCP Datastore support.

I'm able to read data from both Spanner and Datastore even tho they are located in different GCP projects. The problem is when I try to save any entity (or perform a primary key operation which have a transaction behind the scenes) I get the following error:
```java
2022-02-10 21:49:27,317 ERROR  [page.PageSync] Error creating page of id 4324324234234 org.springframework.beans.factory.NoSuchBeanDefinitionException: No bean named 'transactionManager' available: No matching TransactionManager bean found for qualifier 'transactionManager' - neither qualifier match nor bean name match!
	at org.springframework.beans.factory.annotation.BeanFactoryAnnotationUtils.qualifiedBeanOfType(BeanFactoryAnnotationUtils.java:136) ~[spring-beans-5.3.8.jar:5.3.8]
	at org.springframework.beans.factory.annotation.BeanFactoryAnnotationUtils.qualifiedBeanOfType(BeanFactoryAnnotationUtils.java:95) ~[spring-beans-5.3.8.jar:5.3.8]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.determineQualifiedTransactionManager(TransactionAspectSupport.java:515) ~[spring-tx-5.3.8.jar:5.3.8]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.determineTransactionManager(TransactionAspectSupport.java:496) ~[spring-tx-5.3.8.jar:5.3.8]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:342) ~[spring-tx-5.3.8.jar:5.3.8]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-5.3.8.jar:5.3.8]
```

I'm basically reading data from Datastore and saving data into Spanner.

Code throwing error:
```java
   this.pageRepository.save(page);
```

Repository implementation:
```java
@Repository
public interface PageRepository extends JpaRepository<Page, UUID> {}
```

My project configuration (only what matters for this specific error):
```xml
  <properties>
    <!-- Spring -->
    <spring-cloud.version>2020.0.3</spring-cloud.version>
    <spring-cloud-gcp.version>2.0.3</spring-cloud-gcp.version>
    <spring.version>2.5.2</spring.version>
  </properties>

  <!-- I do not use any Spring Parent since I have my own parents -->
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.springframework.cloud</groupId>
        <artifactId>spring-cloud-dependencies</artifactId>
        <version>${spring-cloud.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
      <dependency>
        <groupId>com.google.cloud</groupId>
        <artifactId>spring-cloud-gcp-dependencies</artifactId>
        <version>${spring-cloud-gcp.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
      <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-dependencies</artifactId>
        <version>${spring.version}</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>

  <dependencies>
    <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>spring-cloud-gcp-starter-data-datastore</artifactId>
    </dependency>
    
    <dependency>
      <groupId>org.springframework.boot</groupId>
      <artifactId>spring-boot-starter-data-jpa</artifactId>
    </dependency>
    
    <dependency>
      <groupId>com.google.cloud</groupId>
      <artifactId>google-cloud-spanner-hibernate-dialect</artifactId>
      <version>1.5.2</version>
    </dependency>
  </dependencies>
```

WIth these properties:

```properties
spring.cloud.gcp.datastore.project-id=my-project-id
spring.datasource.url=jdbc:cloudspanner:/projects/other-project-id/instances/my-instance/databases/my-database
spring.datasource.driver-class-name=com.google.cloud.spanner.jdbc.JdbcDriver
spring.jpa.database-platform=com.google.cloud.spanner.hibernate.SpannerDialect
```

I can't use Spanner from Spring Cloud GCP because I already have a project using JPA and now I need to access datastore in the same project to copy data to Spanner.

If I remove the `spring-cloud-gcp-starter-data-datastore` dependency from the project (and comment all codes using datastore) I can save data into the Spanner instance without errors.

### PR DESCRIPTION
**Pull request by bijukunjummen**
Fixes #944 

Conditional for Spanner is based on missing beans of type SpannerTransactionManager instead of PlatformTransactionManager, this way it always gets created if Spanner libs are pulled in 

Conditional for Datastore is based on missing beans of type DatastoreTransactionManager instead of PlatformTransactionManager, this way it gets created if Datastore libs are pulled in.

It is the user's responsibility to designate the transaction's right transaction manager:

```
@Transactional(transactionManager = "spannerTransactionManager")
```

```
@Transactional(transactionManager = "datastoreTransactionManager")
```

### CODE REVIEW

**Review by zhumin8 — CHANGES_REQUESTED** [review]
Thanks for fixing this and sorry for letting this hang for so long!
 
Just one minor request: Do you mind also adding a note to the documentation that user needs to designate the transaction's right transaction manager if multiple transaction managers are brought in?
The corresponding sections in spanner documentation are [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/asciidoc/spanner.adoc?plain=1#L951) and [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/md/spanner.md?plain=1#L1111), datastore documentation are [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/asciidoc/datastore.adoc?plain=1#L849) and [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/md/datastore.md?plain=1#L933). (We are keeping adoc and md copies of documentations temporarily)

**Comment by zhumin8** [comment]
On a separate note, I've verified this solution manually as we do not have a test setup with both Datastore and Spanner. @JoeWang1127 @meltsufin For future, what do you think about setting up an integration test with multiple data modules we support so we can capture bugs like this one? Or alternatively, there is an existing [sample](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/3.x/spring-cloud-gcp-samples/spring-cloud-gcp-data-multi-sample) that we could add a test to.

**Comment by meltsufin** [comment]
I think that re-using the existing sample that already combines Spanner and Datastore makes sense.

Just to clarify, this change doesn't require you to specify `transactionManager` with the `@Transactional`, if only one of the staters is used?

**Comment by zhumin8** [comment]
> Just to clarify, this change doesn't require you to specify `transactionManager` with the `@Transactional`, if only one of the staters is used?

Yes, no change of behavior when only one started used. This is the case with existing IT tests with `@Transactional` in [Spanner](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-data-spanner/src/test/java/com/google/cloud/spring/data/spanner/core/it/SpannerTemplateIntegrationTests.java#L209) and [Datastore](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-data-datastore/src/test/java/com/google/cloud/spring/data/datastore/it/TransactionalTemplateService.java#L34) and they both work as intended. (Integration tests are skipped for this PR because it's from a forked repo, but I've verified this behavior running from my machine)

**Comment by bijukunjummen** [comment]
> Thanks for fixing this and sorry for letting this hang for so long!
> 
> Just one minor request: Do you mind also adding a note to the documentation that user needs to designate the transaction's right transaction manager if multiple transaction managers are brought in? The corresponding sections in spanner documentation are [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/asciidoc/spanner.adoc?plain=1#L951) and [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/md/spanner.md?plain=1#L1111), datastore documentation are [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/asciidoc/datastore.adoc?plain=1#L849) and [here](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/3.x/docs/src/main/md/datastore.md?plain=1#L933). (We are keeping adoc and md copies of documentations temporarily)

Done @zhumin8. Thanks for pointing out the exact location for the documentation. Let me know if you want any of the text changed.

**Comment by JoeWang1127** [comment]
Hi @bijukunjummen, could you add a test in this [sample](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/3.x/spring-cloud-gcp-samples/spring-cloud-gcp-data-multi-sample) to verify it's working when connecting Spanner and Datastore?

**Comment by bijukunjummen** [comment]
> Hi @bijukunjummen, could you add a test in this [sample](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/tree/3.x/spring-cloud-gcp-samples/spring-cloud-gcp-data-multi-sample) to verify it's working when connecting Spanner and Datastore?

Done @JoeWang1127

**Comment by JoeWang1127** [comment]
@bijukunjummen do you think we should change/add the integration tests in samples to verify the samples are working as expected?

**Review by zhumin8 — CHANGES_REQUESTED** [review]
Sample looks great! Simple and minimum additions. 
Since you are at it, do you mind also adding a test similar to [this existing one](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-samples/spring-cloud-gcp-data-multi-sample/src/test/java/com/example/MultipleDataModuleIntegrationTest.java)? It should basically do the same thing as your test, but the test runs in our CI so we are sure this functionality does not accidentally break in the future. 
It could also be useful to mention this change in the sample readme, along the lines of “This sample also demonstrates usage of transactions with both modules.”

File: docs/src/main/asciidoc/datastore.adoc [hunk 1]
```diff
@@ -854,6 +854,13 @@ This feature requires a bean of `DatastoreTransactionManager`, which is provided
 If a method annotated with `@Transactional` calls another method also annotated, then both methods will work within the same transaction.
 `performTransaction` cannot be used in `@Transactional` annotated methods because Cloud Datastore does not support transactions within transactions.
 
+Other Google Cloud database related libraries like spanner, firestore can introduce `PlatformTransactionManager` beans, and can interfere with Datastore Transaction Manager. To disambiguate, explicitly specify the name of the transaction manager bean for such `@Transactional` methods, for eg.
```
Comment (meltsufin) [inline docs/src/main/asciidoc/datastore.adoc:857]: Just a wording suggestion. Please update the other instances as well if you like the rewording.

```suggestion
Other Google Cloud database-related integrations like Spanner and Firestore can introduce `PlatformTransactionManager` beans, and can interfere with the Datastore Transaction Manager. To disambiguate, explicitly specify the name of the transaction manager bean for such `@Transactional` methods. Example:
```

**Review by zhumin8 — APPROVED** [review]
LGTM.
I realized since integration test are not run for this pr, it's fair that adding sample test should not be part of it. I'll add it in a followup pr once this is in.

If you could address the comment on documentation wording, I am good to merge this in.

**Pull request merged**